File: evaluations/OlderFilterTests/TestFilter.cpp

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <cmath>

class IIRFilter
{
public:
    // Constructor for 8th order filter
    IIRFilter(const std::vector<double>& a_coeffs, const std::vector<double>& b_coeffs, const std::vector<double>& zi_initial)
    {
        for (int i = 0; i < 9; i++) {
            a[i] = a_coeffs[i];
            b[i] = b_coeffs[i];
        }
        for (int i = 0; i < 8; i++) {
            zi[i] = zi_initial[i];
            v[i] = zi[i];  // Initialize delayed elements to initial conditions
        }
    }

    ~IIRFilter() {}

    double process(double x1) {
        // Calculate the output using the Direct Form I implementation
        double y1 = b[0] * x1 + v[0];
        for (int i = 0; i < 7; i++) {
            v[i] = b[i + 1] * x1 + v[i + 1] - a[i + 1] * y1;
        }
        v[7] = b[8] * x1 - a[8] * y1;
        return y1;
    }

private:
    double a[9];
    double b[9];
    double zi[8];
    double v[8];
};

std::vector<double> extractColumnFromCSVFile(const std::string& filename, int colIdx) {
    std::vector<double> columnData;
    std::ifstream file(filename);
    std::string line;

    // Check if the file opened successfully
    if (!file.is_open()) {
        std::cerr << "Error opening file: " << filename << std::endl;
        return columnData;
    }

    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string cell;
        int columnIndex = 0;

        while (std::getline(ss, cell, ',')) {
            if (columnIndex == colIdx) {
                try {
                    double value = std::stod(cell);
                    columnData.push_back(value);
                } catch (const std::invalid_argument& e) {
                    std::cerr << "Invalid number in file: " << cell << std::endl;
                } catch (const std::out_of_range& e) {
                    std::cerr << "Number out of range in file: " << cell << std::endl;
                }
            }
            columnIndex++;
        }
    }

    file.close();
    return columnData;
}

int main() {
    // Coefficients from the Python script
    std::vector<double> a = {1.0, -6.05960751, 16.45391545, -26.18801509, 26.74607739, -17.95499326, 7.73746173, -1.9574456, 0.22281157};
    std::vector<double> b = {0.00336282, 0.0, -0.01345126, 0.0, 0.02017689, 0.0, -0.01345126, 0.0, 0.00336282};
    std::vector<double> zi = {-0.00336282, -0.00336282, 0.01008845, 0.01008845, -0.01008845, -0.01008845, 0.00336282, 0.00336282};
    // std::vector<double> zi = {0,0,0,0,0,0,0,0};

    // Create the filter
    IIRFilter filter(a, b, zi);

// #define USE_GENERATED_DATA
#ifdef USE_GENERATED_DATA
    // Generate a sample signal for testing
    double fs = 25.0;
    double t = 1.0;
    int N = static_cast<int>(fs * t);
    std::vector<double> data(N);
    for (int i = 0; i < N; i++) {
        data[i] = sin(1.0 * 2 * M_PI * i / fs) + 0.5 * sin(2.0 * 2 * M_PI * i / fs);
    }
#else
    // Read from CSV file 20240519_1_ADC_Data.csv second column is the Red ADC data
    auto data = extractColumnFromCSVFile("20240519_1_ADC_Data.csv", 1);
    int N = data.size();
#endif

    // Apply the filter
    std::cout << "Filtered data: [ ";
    int col_count = 0;
    for (int i = 0; i < N; i++) {
        double filtered_value = filter.process(data[i]);

        if (i >= 1000 && i <= 2000) {
            printf("x1: %f, y1: %f\n", data[i], filtered_value);
        }
        // std::cout << filtered_value << " ";
        // col_count++;
        // if (col_count == 6) {
        //     std::cout << std::endl;
        //     col_count = 0;
        // }
    }
    std::cout << "]" << std::endl;

    return 0;
}

File: evaluations/OlderFilterTests/TestFilter.py
import numpy as np
from scipy.signal import butter, lfilter_zi, filtfilt, lfilter
import csv

# Filter parameters
fs = 25.0  # Sample rate, Hz
lowcut = 0.75  # Desired low cut frequency of the filter, Hz
highcut = 3.0  # Desired high cut frequency of the filter, Hz
order = 2  # Filter order (this results in an 8th-order filter for bandpass)

# Create Butterworth bandpass filter
b, a = butter(order, [lowcut, highcut], fs=fs, btype='band')

# Print coefficients
print("b coefficients:", b)
print("a coefficients:", a)

# Generate initial conditions for the filter
zi = lfilter_zi(b, a) * 0  # Initial conditions (assuming zero initial conditions)

# Print initial conditions
print("Initial conditions (zi):", zi)

np.set_printoptions(suppress=True, formatter={'float_kind': '{:0.8f}'.format})
USE_GENERATED_DATA = True

if USE_GENERATED_DATA:
    # Generate a sample signal for testing
    duration = 10.0  # seconds
    t = np.linspace(0, duration, int(fs*duration), endpoint=False)  # N seconds of data at 25 Hz
    data = np.sin(1.0 * 2 * np.pi * t) + 0.5 * np.sin(2.0 * 2 * np.pi * t)
else:
    # Read from CSV file 20240519_1_ADC_Data.csv second column is the Red ADC data
    data = []
    with open('20240519_1_ADC_Data.csv', newline='') as csvfile:
        reader = csv.reader(csvfile, delimiter=',')
        for row in reader:
            if row[1] == "Red":
                continue
            data.append(float(row[1]))

# Print data
# print("Data:", data)
    
# Apply the filter
filtered_data = lfilter(b, a, data)  # Using lfilter with initial conditions

zipped = zip(data, filtered_data)
for d, f in zipped:
    print("Data: ", d, "Filtered: ", f)
# print("Filtered data:", filtered_data[0:100])
